android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java

```diff
@@ -0,0 +1,465 @@
+
+import java.beans.PropertyChangeListener;
+import java.beans.PropertyChangeSupport;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+public class ResourcesUpdateTool implements KeyboardEventHandler.OnKeyboardDownloadEventListener, CloudRepository.UpdateHandler{
+
+  /**
+   * Force resource update.
+   * Only for testing, should be false for merging
+   */
+  public static final boolean FORCE_RESOURCE_UPDATE = false;
+
+  /**
+   * Send update notifications.
+   */
+  public static final boolean SEND_UPDATE_NOTIFICATIONS = true;
+
+  private static final class OngoingUpdate
+  {
+    Integer notificationid;
+    String updateid;
+    Bundle bundle;
+  }
+
+  private Context currentContext;
+
+  private boolean updateCheckFailed = false;
+  private boolean updateFailed = false;
+  private Calendar lastUpdateCheck = null;
+  private boolean checkingUpdates = false;
+
+  private LinkedHashMap<String, OngoingUpdate> openUpdates = new LinkedHashMap<>();
+  private int failedUpdateCount = 0;
+
+  private PropertyChangeSupport propertyChangeSupport = new PropertyChangeSupport(this);
+
+
+  private AtomicInteger notificationid = new AtomicInteger(1);
+
+  private void checkResourceUpdatesInternal(boolean anInitialize) {
+
+    Runnable onSuccess = new Runnable() {
+      public void run() {
+        if(! openUpdates.isEmpty()) {
+          return;
+        }
+
+        Toast.makeText(currentContext, currentContext.getString(R.string.update_check_current), Toast.LENGTH_SHORT).show();
+        lastUpdateCheck = Calendar.getInstance();
+        SharedPreferences prefs = currentContext.getSharedPreferences(currentContext.getString(R.string.kma_prefs_name), Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = prefs.edit();
+        editor.putLong("lastUpdateCheck", lastUpdateCheck.getTime().getTime());
+        editor.commit();
+        checkingUpdates = false;
+      }
+    };
+
+    Runnable onFailure = new Runnable() {
+      public void run() {
+        if(! openUpdates.isEmpty()) {
+          return;
+        }
+
+        Toast.makeText(currentContext, currentContext.getString(R.string.update_check_unavailable), Toast.LENGTH_SHORT).show();
+        lastUpdateCheck = Calendar.getInstance();
+        updateCheckFailed = true;
+        checkingUpdates = false;
+      }
+    };
+
+    checkingUpdates = true;
+    if(anInitialize)
+      CloudRepository.shared.initializeDataSet(currentContext, this, onSuccess, onFailure);
+    else
+      CloudRepository.shared.updateDatasetIfNeeded(currentContext, this, onSuccess, onFailure);
+
+  }
+
+  public static void createNotificationChannel(Context aContext) {
+    // Create the NotificationChannel, but only on API 26+ because
+    // the NotificationChannel class is new and not in the support library
```

Comment:
What's the behavior on devices older than Oreo when there's no notification channel?

Reply:
you can send notifications without a channel for old android versions